src/client_side.cc

```diff
@@ -3952,7 +3952,7 @@ ConnStateData::checkLogging()
 
     // do not log connections that closed after a transaction (it is normal)
     // TODO: access_log needs ACLs to match received-no-bytes connections
-    if (pipeline.nrequests && inBuf.isEmpty())
+    if (pipeline.nrequests && !pendingRequestBytes())
```

Comment:
This and similar PR changes effectively assert that inBuf is not the input buffer -- the input buffer is a more complex concept that may, for example, use both inBuf and parser_ storage. I do not think this PR applies that new design consistently. For example, this PR does not update requestErrorBytes code in Http::One::Server::buildHttpRequest() that uses the old "inBuf is the input buffer" design. Another red flag is `preservedClientData = inBuf` assignment in ConnStateData::parseHttpRequest(). I also think that migrating towards that more complex state will probably complicate long-term code improvements and maintenance.

I suspect that instead of making "input buffer" concept more complex and increasing cases where parser_ is used outside of the "current request" context, we should keep inBuf concept intact but create "current request" earlier (as we discussed in another issue) and moving parser_ inside ClientHttpRequest. AFAICT, if we do that, then those changes will fix the bug this PR is fixing (and more).

If you agree, let's block this PR on that "create current request earlier" work and see what happens.

Reply:
> This and similar PR changes effectively assert that inBuf is not the input buffer -- the input buffer is a more complex concept

I was thinking about this PR changes as a minor improvement without changing concepts and making general assertions. The added wrappers basically do not change anything (except for that improvement) - it is clear that the input buffer is indeed a 'complex concept' with all those incremental parsing, variables pointing at the insides of inBuf, etc., and all this existed before and still exists.

> Another red flag is preservedClientData = inBuf assignment

I do not see a red flag here especially if we are going to move checkLogging() into terminateAll() and remove clearPendingRequestBytes(). In this case, we should not care about cleaning up that preservedClientData.

Overall, I see nothing bad with this improvement. The new pendingRequestBytes() method should work correctly - i.e., return true if there are 'pending bytes' (and false otherwise). If it is true - we should not care about other variables in this PR, such as preservedClientData (i.e., copying inBuf or pointing to it). Some other PRs may refactor this, of course.

>  suspect that instead of making "input buffer" concept more complex 

If we end up having only one new pendingRequestBytes() method returning boolean, then the new approach looks simpler IMO. It is easier to check the result of a self-documented method than analyze the state of the input buffer (i.e., obscure !inBuf.empty() call).

> create "current request" earlier

If we do that, we can keep calling pendingRequestBytes() and simply remove parser from its implementation.

Reply:
> > This and similar PR changes effectively assert that inBuf is not the input buffer -- the input buffer is a more complex concept

> it is clear that the input buffer is indeed a 'complex concept' with all those incremental parsing, variables pointing at the insides of inBuf, etc., and all this existed before and still exists.

My claim is not that inBuf (usage) is complex or became more complex. My claim is that this PR asserts that "input buffer is not (just) inBuf" but then does not apply that assertion/principle/concept consistently to code that seemingly violates that newly discovered and codified assertion. This PR probably does not increase inBuf usage complexity, but my concern is different.

> > instead of making "input buffer" concept more complex

Here, I should not have said "making". PR asserts/acknowledges what is _already present_ in official code. We are documenting/exposing that (arguably previously hidden) principle but then are not being consistent about applying that principle to arguably in-scope code that seemingly uses inBuf without thinking about other input sources.

> If we end up having only one new pendingRequestBytes() method returning boolean, then the new approach looks simpler IMO. It is easier to check the result of a self-documented method than analyze the state of the input buffer (i.e., obscure !inBuf.empty() call).

For human developers, it will be "easier" to check inBuf.empty() in one place while calling pendingRequestBytes() in another, and reviewers will probably miss that inconsistency. Similar direct misuse of inBuf will continue to be present throughout the code, causing or preserving subtle bugs.

I am not sure we can address this concern by "creating current request earlier", but if we can, we need to consider doing that.